### PR TITLE
Vam 521 orwdal32 save allergy

### PR DESCRIPTION
**Pull request by kekuso**
- Added the ORWDAL32 SAVE ALLERGY Implementation
- Created additional services for AllergyAssessment create, lookup, update, and delete

### CODE REVIEW

File: emulation/models/orwdal32-save-allergy.js [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        parameterType: 'OBJECT',
+        description: 'Object with Allergy properties',
+        required: true,
+        example: {
+            GMRAERR: 'YES',
+            GMRAERRDT: 'NOW',
+            GMRAERRBY: 'ALEXANDER,ROBERT',
+            GMRAERRCMTS: ['wah, a baaad un!'],
+        },
+        listMemberParameters: [{ // to be converted into an Allergy Object that createAllergy understands
```
Comment (mfuroyama) [inline emulation/models/orwdal32-save-allergy.js]: This is really minor, but we _might_ have to tweak our model validity checks to allow this to pass through.

File: emulation/models/orwdal32-save-allergy.js [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+        // Check if allergyPropertiesInput contains a No Known Allergy property
+        if (allergyPropertiesInput.GMRANKA === 'YES') {
+            // Create or update an Allergy Assessment
+            return allergyService.createNoReactionsAllergyAssessment(session, callback);
+        }
+
+        // Create a new allergy
+        // First create a mapping (Object) of the listMemberParameters array
+        const parameterMetaObj = this.rpc.inputParameters[2].listMemberParameters.reduce((o, param) => {
```
Comment (mfuroyama) [inline emulation/models/orwdal32-save-allergy.js]: This is yet another minor thing, but you might want to consider replacing this with the lodash `_.keyBy` function, which I think does exactly the same thing:
```Javascript
const parameterMetaObj = _.keyBy(this.rpc.inputParameters[2].listMemberParameters, 'key');
```

https://lodash.com/docs/4.17.5#keyBy

File: emulation/models/tests/orwdal32-save-allergy-spec.js [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    let rpc;
+    let context;
+
+    beforeAll(() => {
+        const res = unitTestUtils.beforeAllScript(rpcEmulatorModel);
+        context = res.context;
+        rpc = res;
+    });
+
+    // invokeRPC and related specs
```
Comment (mfuroyama) [inline emulation/models/tests/orwdal32-save-allergy-spec.js]: You might want to consider bolstering the test cases here with logic that checks any field calculation done in `invokeRPC` (e.g. test cases that use `GMRAGNT` or say, have additional comments)

File: services/allergyService/createAllergy.js [hunk 4]
```diff
@@ -14,20 +14,28 @@ const REQUIRED_INPUT_PROPERTIES = [
...
- * @param {String} args.allergyType Type of allergy. Possible values: DRUG, DRUG, FOOD, FOOD, OTHER.
- * @param {String} args.mechanism Allergy mechanism. Possible values: ALLERGY, PHARMACOLOGI", UNKNOWN.
+ * @param {String} args.allergyType Type of allergy. Possible values: DRUG, FOOD, (DRUG, FOOD), OTHER.
+ * @param {String} args.mechanism Allergy mechanism. Possible values: ALLERGY, PHARMACOLOGIC, UNKNOWN.
  * @param {String} args.observedOrHistorical Observed or historical indicator. Possible values: OBSERVED, HISTORICAL.
  * @param {String} args.allergySeverity Allergy severity. Possible values: MILD, MODERATE, SEVERE.
  * @param {Date} args.dateOccurred Date occurred value.
  * @param {Array} args.reactions List of reaction identifiers.
  * @param {Array} args.comments List of allergy comments.
- * @param {String} args.enteredBy Entered by identifier. Defaults to user.
```
Comment (mfuroyama) [inline services/allergyService/createAllergy.js]: 👍 **THANK YOU** for doing this!!

File: services/allergyService/createAllergy.js [hunk 5]
```diff
@@ -47,12 +55,12 @@ function createAllergy(args, context, callback) {
...
         return true;
     });
-    if (!validInput) return callback(null, { failure: `Failed to create Allergy (${requiredProp} required)` });
+    if (!validInput) return callback(null, { error: `Failed to create Allergy (${requiredProp} required)` });
 
     // Create a new Allergy Object as specified by the arguments
     const allergyObj = _.pick(args, PROPERTIES);
-    allergyObj.dateOccurred = moment(args.dateOccured).format('YYYY-MM-DD');
-    allergyObj.enteredBy = _.get(args, 'enteredBy', { id: context.userId, label: context.userLabel });
+    allergyObj.dateOccurred = moment(args.dateOccurred).format('YYYY-MM-DD');
```
Comment (mfuroyama) [inline services/allergyService/createAllergy.js]: This is fine, but for fields in the MongoDB with a 'DATE' type in the schema, you can just pass in an ISO date string (i.e. `var d = new Date(); d.toISOString();`) and the Mongoose layer takes care of converting the string into a Date object for storage. No real need for the `moment` conversion here

**Comment by mfuroyama** [comment]
Other than the beefing-up of the unit tests, this looks awesome to me, man! This was a lot of hard work...thanks for coming through!

File: emulation/models/orwdal32-save-allergy.js (hunk 2 above)
Comment (kekuso) [reply emulation/models/orwdal32-save-allergy.js]: beautiful 👏

File: services/allergyService/createNoReactionsAllergyAssessment.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        service.find('AllergyAssessment', {
+            query: { 'patient.id': context.patientId },
+            projection: serviceUtils.createProjection(['id', 'hasReactions']),
+        }, cb);
+    };
+
+    const processAllergyAssessment = (res, cb) => {
+        // if no assessment exists for the current patient, create a new assessment with no known allergies (hasReactions: false)
+        if (!res || !res.length) {
+            const newAllergyAssObj = {
```
Comment (mfuroyama) [inline services/allergyService/createNoReactionsAllergyAssessment.js]: Nice allergy ass! LOL!

I think the only thing left here is to make sure that you create the allergy assessment object with a URN that contains the IEN of the current patient. If you merge the latest master into this branch, you should pick up the changes that let you freely specify the `id` field.

**Pull request merged**